app/services/sso/openid_connect.rb

```diff
@@ -0,0 +1,56 @@
+module SSO
+  class OpenidConnect < Base
+    delegate :session, :to => :controller
+    attr_reader :current_user
+
+    def available?
+      controller.api_request? && jwt_token_set?
+    end
+
+    def authenticate!
+      payload = jwt_token.decode
+      user = create_user_from_jwt(payload)
+      @current_user = user
+      store(payload)
+      user&.login
+    rescue JWT::DecodeError
+      Rails.logger.error "JWT SSO: Failed to decode JWT."
+      nil
+    end
+
+    def authenticated?
+      self.user = User.current.presence || authenticate!
+    end
+
+    private
+
+    def jwt_token
+      @jwt_token ||= jwt_token_from_request
+    end
+
+    def jwt_token_from_request
+      token = request.authorization.split(' ')[1]
+      OidcJwtToken.new(token)
+    end
+
+    def jwt_token_set?
+      request.authorization.present? && request.authorization.start_with?('Bearer')
+    end
+
+    def store(payload)
+      session[:sso_method] = self.class.to_s
+      session[:expires_at] = payload["exp"]
+    end
+
+    def create_user_from_jwt(payload)
+      User.find_or_create_external_user(
+        { login: payload['preferred_username'],
+          mail: payload['email'],
+          firstname: payload['given_name'],
+          lastname: payload['family_name']},
+        Setting['authorize_login_delegation_auth_source_user_autocreate']
+      )
+      User.unscoped.except_hidden.find_by_login(payload['preferred_username'])
```

Comment:
Can we change the implementation of `find_or_create_external_method` to either return the object or nil (which is also falsy - so the change should be transparent)?

Reply:
I agree with the suggestion. However, I feel this change should be handled in a separate PR since it is an existing function being used in [apache.rb](https://github.com/theforeman/foreman/blob/develop/app/services/sso/apache.rb#L38) too and hence will require more analysis. What do you think?

Reply:
I'd just change the implementation in this PR.